src/bootstrap/native.rs

```diff
@@ -321,6 +310,38 @@ impl Step for Llvm {
 
         cfg.build();
 
+        // On macOS, we want the build directory in LLVM to match linux with an
+        // unversioned libLLVM.dylib and versioned
+        // `libLLVM-$MAJOR_VERSION-$VERSION_SUFFIX.dylib`. This is necessary for
+        // llvm-config to work, for example. Unfortunately, LLVM itself does not
+        // currently produce the versioned library name despite llvm-config
+        // expecting it, so for now just manually create it.
+        if target.contains("apple-darwin") && builder.config.llvm_link_shared {
+            let libdir = out_dir.join("lib");
+            let llvm_dylib_path = libdir.join("libLLVM.dylib");
+            if let Some(suffix) = crate::native::llvm_version_suffix(builder) {
+                let mut cmd = Command::new(&build_llvm_config);
+                let version = output(cmd.arg("--version"));
+                let mut parts = version.split('.').take(2).filter_map(|s| s.parse::<u32>().ok());
+                if let Some(major) = parts.next() {
+                    builder.copy(
+                        &llvm_dylib_path,
+                        &libdir.join(&format!("libLLVM-{}{}.dylib", major, suffix)),
```

Comment:
Can you explain a bit more what depends on this? How would a binary depend on a library that doesn't exist as part of the build process?

Reply:
llvm-config when run with --link-shared will look for this dylib and report an error if it doesn't exist (https://github.com/rust-lang/rust/pull/76810#issuecomment-693795808 has a snapshot of what the error looks like). That error isn't really accurate in the sense that the linker will resolve the `-lLLVM-11-rust-dev` it gets passed to libLLVM.dylib happily in my testing. I don't really understand quite how that works, though I suspect some of the macOS-specific install_name_tool logic.

I think it's plausible that we could just "touch" this file here rather than copying over the LLVM dylib, but I suspect that somewhere down the line something will want to read pieces of this file... (I haven't tested just touching it).